# Importing Necessary Libraries

In [35]:
# libraries for data manipulation
import numpy as np
import pandas as pd
import re

# libraries for EDA and data visualization
import matplotlib.pyplot as plt
import seaborn as sns

# libraries for model builing and evaluation
import sklearn
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import BernoulliNB, GaussianNB
from sklearn.ensemble import AdaBoostClassifier, RandomForestClassifier
from sklearn.metrics import accuracy_score, auc, confusion_matrix, roc_auc_score, recall_score, precision_score, precision_recall_curve, mean_squared_error

import xgboost
from xgboost import XGBClassifier

# other libraries
import warnings
warnings.filterwarnings('ignore')

# Reading and Understanding data

In [36]:
data_train = pd.read_csv('train.csv')
data_test = pd.read_csv('test.csv')
data_dict = pd.read_csv('data_dictionary.csv')

In [37]:
data_dict

Acronyms                                        Description
0     CIRCLE_ID  Telecom circle area to which the customer belo...
1           LOC            Local calls  within same telecom circle
2           STD              STD calls  outside the calling circle
3            IC                                     Incoming calls
4            OG                                     Outgoing calls
5           T2T  Operator T to T ie within same operator mobile...
6       T2M                    Operator T to other operator mobile
7       T2O                Operator T to other operator fixed line
8       T2F                         Operator T to fixed lines of T
9       T2C                      Operator T to its own call center
10     ARPU                               Average revenue per user
11      MOU                          Minutes of usage  voice calls
12      AON      Age on network  number of days the customer is...
13     ONNET     All kind of calls within the same operator net...
14   OFFNET       All kind of calls outside the operator T network
15         ROAM  Indicates that customer is in roaming zone dur...
16       SPL                                         Special calls
17      ISD                                              ISD calls
18     RECH                                               Recharge
19      NUM                                                 Number
20      AMT                               Amount in local currency
21      MAX                                                Maximum
22     DATA                                        Mobile internet
23       3G                                              G network
24       AV                                                Average
25      VOL                     Mobile internet usage volume in MB
26       2G                                              G network
27      PCK                  Prepaid service schemes called  PACKS
28    NIGHT         Scheme to use during specific night hours only
29  MONTHLY      Service schemes with validity equivalent to a ...
30    SACHET     Service schemes with validity smaller than a m...
31      *.6                              KPI for the month of June
32      *.7                              KPI for the month of July
33      *.8                            KPI for the month of August
34      FB_USER  Service scheme to avail services of Facebook a...
35      VBC      Volume based cost  when no specific scheme is ...

In [38]:
data_train.head()

id  circle_id  loc_og_t2o_mou  ...  jul_vbc_3g  jun_vbc_3g churn_probability
0   0        109             0.0  ...         0.0         0.0                 0
1   1        109             0.0  ...         0.0         0.0                 0
2   2        109             0.0  ...         0.0         0.0                 0
3   3        109             0.0  ...         0.0         0.0                 0
4   4        109             0.0  ...         0.0         0.0                 0

[5 rows x 172 columns]

In [39]:
data_test.head()

id  circle_id  loc_og_t2o_mou  ...  aug_vbc_3g  jul_vbc_3g jun_vbc_3g
0  69999        109             0.0  ...        0.00        0.00       0.00
1  70000        109             0.0  ...        0.00        0.00       0.00
2  70001        109             0.0  ...      525.61      758.41     241.84
3  70002        109             0.0  ...        0.00        0.00       0.00
4  70003        109             0.0  ...        0.00        0.00       0.00

[5 rows x 171 columns]

In [40]:
data_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Columns: 172 entries, id to churn_probability
dtypes: float64(135), int64(28), object(9)
memory usage: 91.9+ MB


In [41]:
data_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Columns: 171 entries, id to jun_vbc_3g
dtypes: float64(135), int64(27), object(9)
memory usage: 39.1+ MB


In [42]:
len(data_train['circle_id'].unique())

1

- Dropping the column **circle_id** as it seems to have a single entry in the entire train dataset.
- Removing the **id** column as it is unnecessary for predictive analysis.

In [43]:
data_train.pop('circle_id')

0        109
1        109
2        109
3        109
4        109
        ... 
69994    109
69995    109
69996    109
69997    109
69998    109
Name: circle_id, Length: 69999, dtype: int64

In [44]:
data_train.pop('id')

0            0
1            1
2            2
3            3
4            4
         ...  
69994    69994
69995    69995
69996    69996
69997    69997
69998    69998
Name: id, Length: 69999, dtype: int64

# Data Cleaning and Preparation

Performing primary feature elimination by keeping only relevant features. Columns having total, max, average and count are kept along with certain other business necessary columns

In [45]:
total_amounts = [i for i in list(data_train.columns) if re.search('total',i)]
avg_amounts = [i for i in list(data_train.columns) if re.search('av',i)]
max_amounts = [i for i in list(data_train.columns) if re.search('max',i)]
count_amounts = [i for i in list(data_train.columns) if re.search('count',i)]
offnetwork_minutes = [i for i in list(data_train.columns) if re.search('offnet',i)]
average_revenue_3g = [i for i in list(data_train.columns) if re.search('arpu.+3g',i)]
average_revenue_2g = [i for i in list(data_train.columns) if re.search('arpu.+2g',i)]
volume_3g = [i for i in list(data_train.columns) if re.search('vol.+3g',i)]
volume_2g = [i for i in list(data_train.columns) if re.search('vol.+2g',i)]
age_on_network = [i for i in list(data_train.columns) if re.search('aon',i)]

In [46]:
variables = [*total_amounts,
             *avg_amounts,
             *max_amounts,
             *count_amounts,
             *offnetwork_minutes, 
             *average_revenue_3g, 
             *average_revenue_2g,
             *volume_3g,
             *volume_2g,
             *age_on_network, 
             'churn_probability']

In [47]:
data_train = data_train[variables]
data_train.head()

total_og_mou_6  total_og_mou_7  ...   aon  churn_probability
0           81.21          221.68  ...  1958                  0
1            0.00           30.73  ...   710                  0
2           56.49           99.36  ...   882                  0
3           76.03           95.98  ...   982                  0
4           63.26           42.94  ...   647                  0

[5 rows x 47 columns]

In [48]:
data_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Data columns (total 47 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   total_og_mou_6      69999 non-null  float64
 1   total_og_mou_7      69999 non-null  float64
 2   total_og_mou_8      69999 non-null  float64
 3   total_ic_mou_6      69999 non-null  float64
 4   total_ic_mou_7      69999 non-null  float64
 5   total_ic_mou_8      69999 non-null  float64
 6   total_rech_num_6    69999 non-null  int64  
 7   total_rech_num_7    69999 non-null  int64  
 8   total_rech_num_8    69999 non-null  int64  
 9   total_rech_amt_6    69999 non-null  int64  
 10  total_rech_amt_7    69999 non-null  int64  
 11  total_rech_amt_8    69999 non-null  int64  
 12  total_rech_data_6   17568 non-null  float64
 13  total_rech_data_7   17865 non-null  float64
 14  total_rech_data_8   18417 non-null  float64
 15  av_rech_amt_data_6  17568 non-null  float64
 16  av_r

## Missing Data Treatment

In [49]:
data_train.isna().sum()/len(data_train)*100

total_og_mou_6         0.000000
total_og_mou_7         0.000000
total_og_mou_8         0.000000
total_ic_mou_6         0.000000
total_ic_mou_7         0.000000
total_ic_mou_8         0.000000
total_rech_num_6       0.000000
total_rech_num_7       0.000000
total_rech_num_8       0.000000
total_rech_amt_6       0.000000
total_rech_amt_7       0.000000
total_rech_amt_8       0.000000
total_rech_data_6     74.902499
total_rech_data_7     74.478207
total_rech_data_8     73.689624
av_rech_amt_data_6    74.902499
av_rech_amt_data_7    74.478207
av_rech_amt_data_8    73.689624
max_rech_amt_6         0.000000
max_rech_amt_7         0.000000
max_rech_amt_8         0.000000
max_rech_data_6       74.902499
max_rech_data_7       74.478207
max_rech_data_8       73.689624
count_rech_2g_6       74.902499
count_rech_2g_7       74.478207
count_rech_2g_8       73.689624
count_rech_3g_6       74.902499
count_rech_3g_7       74.478207
count_rech_3g_8       73.689624
offnet_mou_6           3.954342
offnet_m

We see that the following columns have more than **70%** data missing. Hence removing the columns from the dataset:
- **total_rech_data_6**
- **total_rech_data_6**
- **total_rech_data_7**
- **total_rech_data_8**
- **av_rech_amt_data_6**
- **av_rech_amt_data_7**
- **av_rech_amt_data_8**
- **max_rech_data_6**
- **max_rech_data_7**
- **max_rech_data_8**
- **count_rech_2g_6**
- **count_rech_2g_7**
- **count_rech_2g_8**
- **count_rech_3g_6**
- **count_rech_3g_7**
- **count_rech_3g_8**
- **arpu_3g_6**
- **arpu_3g_7**
- **arpu_3g_8**
- **arpu_2g_6**
- **arpu_2g_7**
- **arpu_2g_8** 

In [51]:
data_train.drop(columns=['total_rech_data_6', 'total_rech_data_6', 'total_rech_data_7', 'total_rech_data_8', 
                         'av_rech_amt_data_6','av_rech_amt_data_7', 'av_rech_amt_data_8', 'max_rech_data_6',
                         'max_rech_data_7', 'max_rech_data_8','count_rech_2g_6', 'count_rech_2g_7', 'count_rech_2g_8', 
                         'count_rech_3g_6', 'count_rech_3g_7','count_rech_3g_8', 'arpu_3g_6', 'arpu_3g_7', 
                         'arpu_3g_8', 'arpu_2g_6', 'arpu_2g_7','arpu_2g_8'], inplace=True)

In [60]:
data_train.isna().sum()/len(data_train)*100

total_og_mou_6       0.000000
total_og_mou_7       0.000000
total_og_mou_8       0.000000
total_ic_mou_6       0.000000
total_ic_mou_7       0.000000
total_ic_mou_8       0.000000
total_rech_num_6     0.000000
total_rech_num_7     0.000000
total_rech_num_8     0.000000
total_rech_amt_6     0.000000
total_rech_amt_7     0.000000
total_rech_amt_8     0.000000
max_rech_amt_6       0.000000
max_rech_amt_7       0.000000
max_rech_amt_8       0.000000
offnet_mou_6         3.954342
offnet_mou_7         3.838626
offnet_mou_8         5.290076
vol_3g_mb_6          0.000000
vol_3g_mb_7          0.000000
vol_3g_mb_8          0.000000
vol_2g_mb_6          0.000000
vol_2g_mb_7          0.000000
vol_2g_mb_8          0.000000
aon                  0.000000
churn_probability    0.000000
dtype: float64

In [61]:
data_train.describe()

total_og_mou_6  total_og_mou_7  ...           aon  churn_probability
count    69999.000000    69999.000000  ...  69999.000000       69999.000000
mean       306.451436      310.572674  ...   1220.639709           0.101887
std        465.502866      479.131770  ...    952.426321           0.302502
min          0.000000        0.000000  ...    180.000000           0.000000
25%         44.780000       42.910000  ...    468.000000           0.000000
50%        145.280000      141.230000  ...    868.000000           0.000000
75%        374.305000      380.045000  ...   1813.000000           0.000000
max      10674.030000     8285.640000  ...   4337.000000           1.000000

[8 rows x 26 columns]

The following columns have less than 5% missing data:
- **offnet_mou_6         3.954342**
- **offnet_mou_7         3.838626**
- **offnet_mou_8         5.290076**

We also see that the above columns are highly right skewed. Hence imputing the columns with their median value.

In [67]:
columns = ['offnet_mou_7', 'offnet_mou_6', 'offnet_mou_8']

for column in columns:
    data_train[column].fillna(data_train[column].median(), inplace=True)

In [69]:
data_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   total_og_mou_6     69999 non-null  float64
 1   total_og_mou_7     69999 non-null  float64
 2   total_og_mou_8     69999 non-null  float64
 3   total_ic_mou_6     69999 non-null  float64
 4   total_ic_mou_7     69999 non-null  float64
 5   total_ic_mou_8     69999 non-null  float64
 6   total_rech_num_6   69999 non-null  int64  
 7   total_rech_num_7   69999 non-null  int64  
 8   total_rech_num_8   69999 non-null  int64  
 9   total_rech_amt_6   69999 non-null  int64  
 10  total_rech_amt_7   69999 non-null  int64  
 11  total_rech_amt_8   69999 non-null  int64  
 12  max_rech_amt_6     69999 non-null  int64  
 13  max_rech_amt_7     69999 non-null  int64  
 14  max_rech_amt_8     69999 non-null  int64  
 15  offnet_mou_6       69999 non-null  float64
 16  offnet_mou_7       699